### PR TITLE
Background Preprocessing

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
The fabled background preprocessing is here. You can enable it by setting `--num-workers` to a value greater than 0!

Some rough estimates of performance increases:
- Fine tuning BlenderBot: 5-10%
- Fine tuning BART: 15-25%
- Training a Bart-sized model on CommonCrawl-like data: 110% speedup (2.1x faster)

**Testing steps**
Sooooo much manual testing. There is also new CI.

### CODE REVIEW

File: parlai/core/teachers.py [hunk 1]
```diff
@@ -2377,32 +2416,46 @@ def _enqueue_request(self):
...
         Queue a request for loading to the data loader.
         """
-        self.data_loader.request_load(self.receive_data, self.get_chunk, ())
+        if self.threading:
+            self.data_loader.request_load(self.receive_data, self.get_chunk, ())
+        else:
+            self.receive_data(self.get_chunk(), direct_result=True)
 
-    def receive_data(self, future):
+    def receive_data(self, future, direct_result=False):
```
Comment (EricMichaelSmith) [inline parlai/core/teachers.py:2424]: Hmm maybe it'd be worth adding a line of documentation on `future`? I understand from the line above that it's a chunk, but offhand I don't even know what a chunk is...

File: parlai/utils/bpe.py [hunk 2]
```diff
@@ -535,15 +541,10 @@ def __init__(self, opt: Opt, shared: TShared = None):
         self.byte_decoder = {v: k for k, v in self.byte_encoder.items()}
         self.bpe_ranks = dict(zip(bpe_merges, range(len(bpe_merges))))
 
-        try:
-            import regex as re
-
-            self.re = re
-        except ImportError:
+        if regex is None:
```
Comment (EricMichaelSmith) [inline parlai/utils/bpe.py:544]: Yeah, this was always weird

**Review by EricMichaelSmith — COMMENTED** [review]
Cool! Makes sense to me - will defer to Kurt to approve because I think he has more context on this but happy to do so if desired. What value of `--num-workers` would you recommend for different cases? How much additional mem does this typically use?

**Comment by stephenroller** [comment]
Number of workers gives diminishing returns, but I have still found that 8 is faster than 4. 10 is the "max" we can do on our cluster based on CPUs-per-GPU.

**Comment by stephenroller** [comment]
As far as memory, it's loading the dataset separately per worker. My chunk teachers haven't had issues tho.

File: parlai/core/teachers.py [hunk 3]
```diff
@@ -2277,12 +2304,23 @@ def __init__(self, opt, shared=None):
                 c for c in self.fold_chunks if c % self.dws == self.rank
             ]
 
+        # deal with --num-workers
+        self.threading = not (opt.get('num_workers', 0) > 0 and self.is_train)
```
Comment (klshuster) [inline parlai/core/teachers.py:2308]: so we still thread during validation?

File: parlai/core/teachers.py [hunk 4]
```diff
@@ -2377,32 +2416,46 @@ def _enqueue_request(self):
...
 
-    def receive_data(self, future):
+    def receive_data(self, future, direct_result=False):
         """
         Loads data.
 
         Load data into self.samples until buffersize is reached.
         """
-        chunk_output, chunk_reset_cnt = future.result()
+        output = future if direct_result else future.result()
```
Comment (klshuster) [inline parlai/core/teachers.py:2430]: nit: `direct_result` is a bit ambiguous --> if true, we return the `future`, not the `future.result()`, which is somewhat counterintuitive?

File: parlai/core/torch_agent.py [hunk 5]
```diff
@@ -162,20 +165,26 @@ def to(self, dev):
         """
         Move all tensors in the batch to a device.
 
+        NOT in place.
```
Comment (klshuster) [inline parlai/core/torch_agent.py:168]: just curious, why not?

File: parlai/core/worlds.py [hunk 6]
```diff
@@ -1210,6 +1218,115 @@ def report(self):
...
 
+class BackgroundDriverWorld(World):
+    def __init__(self, opt: Opt, world: World):
+        self.world = world
+        super().__init__(opt, agents=world.agents, shared=None)
+
+        import torch.multiprocessing as mp
+
+        self._num_workers = self.opt['num_workers']
+        self._process_queue = mp.Queue(maxsize=4 * self._num_workers)
```
Comment (klshuster) [inline parlai/core/worlds.py:1229]: arbitrary?

File: parlai/core/worlds.py [hunk 7]
```diff
@@ -1210,6 +1218,115 @@ def report(self):
...
+        for p in self._process_pool.processes:
+            p.kill()
+
+
+class BackgroundWorkerDynamicBatchWorld(DynamicBatchWorld):
+    @classmethod
+    def launch_process(cls, index, opt, model_agent, process_queue):
+        import torch
+
+        torch.set_num_threads(1)  # prevent threads from spawning in this worker
```
Comment (klshuster) [inline parlai/core/worlds.py:1300]: is this local in scope?

File: parlai/core/worlds.py [hunk 8]
```diff
@@ -1210,6 +1218,115 @@ def report(self):
...
+
+    def __del__(self):
+        logging.debug("Killing all the worker processes")
+        for p in self._process_pool.processes:
+            p.kill()
+
+
+class BackgroundWorkerDynamicBatchWorld(DynamicBatchWorld):
+    @classmethod
+    def launch_process(cls, index, opt, model_agent, process_queue):
```
Comment (klshuster) [inline parlai/core/worlds.py:1297]: oh is index the implicit first arg for `mp.start_processes`?

File: parlai/core/worlds.py (hunk 6 above)
Comment (stephenroller) [reply parlai/core/worlds.py:1229]: ya seemed good to me lol

File: parlai/core/worlds.py (hunk 7 above)
Comment (stephenroller) [reply parlai/core/worlds.py:1300]: ya i think so but not 100% sure

File: parlai/core/worlds.py (hunk 8 above)
Comment (stephenroller) [reply parlai/core/worlds.py:1297]: yes

File: parlai/core/teachers.py (hunk 4 above)
Comment (stephenroller) [reply parlai/core/teachers.py:2430]: Okay, refactored to be more clear I hope

File: parlai/core/teachers.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/teachers.py:2319]: Added some clarifying comments

File: parlai/core/worlds.py (hunk 8 above)
Comment (stephenroller) [reply parlai/core/worlds.py:1309]: added a comment

File: parlai/core/worlds.py (hunk 6 above)
Comment (stephenroller) [reply parlai/core/worlds.py:1234]: left a comment

File: docs/source/tutorial_fast.md [hunk 9]
```diff
@@ -2,50 +2,65 @@
...
+runs in ParlAI. These tricks tend to work best with generative models, but some
+can also be used with others.
+
+:::{warning} Retuning-hyperparameters
+Many of these options are effectively tricks for changing the batchsize.
+Note that you will want to freshly tune your learning rate when using
+these tricks, or simply use these options in the first place.
+
+This tutorial is only for illustration purposes of how to speed up training,
+and may not get you the best _performing_ model.
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:15]: assuming this is just a standard disclaimer? or d you actually see worse performance?

File: docs/source/tutorial_fast.md [hunk 10]
```diff
@@ -2,50 +2,65 @@
...
+:::
 
 A summary of the speedups is in this table:
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:18]: maybe indicate here, summary of speedups "for 90m blenderbot model" or something like that

File: docs/source/tutorial_fast.md [hunk 11]
```diff
@@ -2,50 +2,65 @@
...
-| Using 4 GPUs            |   47s |   3s |   50s |   11.0x |
+| Method                   | Train | Eval | Total | Speedup |
+| ------------------------ | ----: | ---: | ----: | ------: |
+| Baseline                 |  579s |  60s |  639s |   1.00x |
+| Skip generation          |  579s |  20s |  599s |   1.07x |
+| Dynamic batching         |  289s |  18s |  307s |   2.08x |
+| FP16                     |  212s |  11s |  223s |   2.60x |
+| Larger batchsize (FP16)  |  168s |  10s |  178s |   3.59x |
+| Background preprocessing |  144s |  10s |  154s |   4.15x |
+| Using 4 GPUs             |   63s |   4s |   67s |   8.64x |
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:28]: gimme 8x all day

File: docs/source/tutorial_fast.md [hunk 12]
```diff
@@ -105,23 +120,23 @@ CUDA cores. We will slightly adjust the size of the network to support this.
...
-    parlai train --dict-file dictfile --model transformer/generator --task convai2 --num-epochs 1.0 \
-        --embedding-size 256 --ffn-size 1024 --n-layers 8 --optimizer adam --learningrate 1e-3 \
+    parlai train --dict-file dictfile --model transformer/generator --task \
+        --task convai2 --num-epochs 1.0 --batchsize 64 --n-layers 8  \
+        --embedding-size 256 --ffn-size 1024 --optimizer adam --learningrate 1e-3 \
         --skip-generation true \
-        --eval-batchsize 256 \
         --dynamic-batching full \
-        --fp16 true --batchsize 128
+        --fp16 true --fp16-impl mem_efficient --batchsize 128 --eval-batchsize 256
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:139]: do we need to explicitly specify which `--fp16-impl` for max efficiency?

File: docs/source/tutorial_fast.md [hunk 13]
```diff
@@ -132,22 +147,48 @@ Without a GPU with FP16 CUDA cores, you may find that FP16 actually slows
 your program. You may still see a benefit from the reduced memory usage though.
 :::
 
+## Background preprocessing
+
+We can further speed things up by doing some preprocessing, such as
+tokenization and dialogue state tracking, in the background thread. This can be
+enabled by setting `--num-workers` to a value greater than 0. A good rule of
+thumb is to set `--num-workers` to the number of CPU cores you have PER GPU.
+Sometimes you can go a bit over, but you will need to play. On my server, there
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:156]: nit: "you will need to play..." "around with it"?

File: docs/source/tutorial_fast.md [hunk 14]
```diff
@@ -132,22 +147,48 @@ Without a GPU with FP16 CUDA cores, you may find that FP16 actually slows
...
+are 4 cores per GPU, so I will set it to 4.
+
+    parlai train --dict-file dictfile --model transformer/generator --task \
+        --task convai2 --num-epochs 1.0 --batchsize 64 --n-layers 8  \
+        --embedding-size 256 --ffn-size 1024 --optimizer adam --learningrate 1e-3 \
+        --skip-generation true \
+        --dynamic-batching full \
+        --fp16 true --fp16-impl mem_efficient --batchsize 128 --eval-batchsize 256 \
+        --num-workers 8
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:165]: nit: you set it to 8 tho specify ur setting to 4

File: docs/source/tutorial_fast.md [hunk 15]
```diff
@@ -132,22 +147,48 @@ Without a GPU with FP16 CUDA cores, you may find that FP16 actually slows
...
+
+Note that we leave batchsize the same: we use the batchsize PER GPU. We expect
+the batchsize will be effectively 4x'd by using 4 GPUs.
+
+Note that launching a multiprocessing train can take time to "warm up" and it
+may take time for you to realize speed improvements when using small, toy
+training runs like in this document. If we train for 4 epochs instead, our FP16
+large-batch run takes 614 seconds and the multi-GPU training takes 222 seconds.
+Also, if you have access to a SLURM cluster, distributed_train is sometimes
+faster than multiprocessin_train. With SLURM, multi-GPU training takes 167 seconds
```
Comment (klshuster) [inline docs/source/tutorial_fast.md:190]: nit: multiprocessin

File: parlai/core/teachers.py (hunk 3 above)
Comment (klshuster) [reply parlai/core/teachers.py:2319]: thanks for the clarification

**Review by klshuster — APPROVED** [review]
really appreciate the clarifications, sorry for the super delayed re-review

File: docs/source/tutorial_fast.md (hunk 9 above)
Comment (stephenroller) [reply docs/source/tutorial_fast.md:15]: The LR isn't optimized for this, and many of them don't train for remotely long enough.

**Comment by sjscotti** [comment]
Hi
I am finetuning BlenderBot2 (400M) model with some domain-specific data and to speed things up I tried to use the `--num-workers` flag on a Windows 10 machine where I am running in an Anaconda environment.  I tries values of `4` and of `8` (I have an 8 core machine) but both gave me errors like shown below.  The training works OK without this flag.  Is there an issue with using this flag in Windows?
Thanks!

```
<prior output is not shown>
...
14:48:29 | creating task(s): jsonfile
14:48:29 | [loading data from json file into task:new_custom_task.txt]
14:48:29 | Metadata does not exist. Please double check your datapath.
Traceback (most recent call last):
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\Scripts\parlai.exe\__main__.py", line 7, in <module>
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\__main__.py", line 14, in main
    superscript_main()
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\core\script.py", line 325, in superscript_main
    return SCRIPT_REGISTRY[cmd].klass._run_from_parser_and_opt(opt, parser)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\core\script.py", line 108, in _run_from_parser_and_opt
    return script.run()
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\scripts\train_model.py", line 932, in run
    self.train_loop = TrainLoop(self.opt)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\scripts\train_model.py", line 349, in __init__
    self.world = create_task(opt, self.agent)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\core\worlds.py", line 1444, in create_task
    world = BackgroundDriverWorld(opt, world)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\core\worlds.py", line 1246, in __init__
    self._process_pool = self._start_processes()
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\parlai\core\worlds.py", line 1254, in _start_processes
    return mp.start_processes(
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\site-packages\torch\multiprocessing\spawn.py", line 169, in start_processes
    mp = multiprocessing.get_context(start_method)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\multiprocessing\context.py", line 239, in get_context
    return super().get_context(method)
  File "C:\Users\Steve\anaconda3\envs\ParlAI\lib\multiprocessing\context.py", line 193, in get_context
    raise ValueError('cannot find context for %r' % method) from None
ValueError: cannot find context for 'fork'
```

**Comment by stephenroller** [comment]
Sorry, we don't support Windows at this time. You can try without bg preprocessing.

**Comment by sjscotti** [comment]
Thanks for the speedy reply @stephenroller 
I am using about 3GB of "shared" GPU memory right now, and I was hoping that background processing would off load enough of the work so that my GPU only used the 8GB of memory on the GPU card.   Do you think that the changes to support Windows would be major?  Any suggestions for what I would need to do to give it a try?
Thanks again!

**Comment by stephenroller** [comment]
I don't think Bg preprocessing will do anything to save you GPU memory. It just does tokenization etc in a background thread so the GPU is always being fed examples.

Unfortunately, our implementation heavily depends on the semantics of fork, and won't work with others.

**Comment by sjscotti** [comment]
Ok, thanks!  I’ve tried all the other tricks I can find to reduce GPU memory such as blocklength  set to 1, fp16 using mem_efficient, and adafactor optimization (I didn’t want to reduce the various truncation settings from the BlenderBot2 settings since I am not sure whether they would affect results from executing the BlenderBot2), so I guess I’ll have to live with the training speed I’ve got.

**Comment by stephenroller** [comment]
If you have multiple GPUs, --ddp-backend zero2 can help.

I've got an intern who is also lamenting how slow training similar models is. I'll let you know how we speed it up.

**Comment by sjscotti** [comment]
Thanks, any help is appreciated.  And unfortunately I only have 1 GPU …
